src/lib/Microsoft.Health.Fhir.Ingest/Service/MeasurementFhirImportService.cs

```diff
@@ -90,9 +120,85 @@ public async Task ProcessEventsAsync(IEnumerable<IEventMessage> events, string t
                 }
             }
 
-            // Step 3: Transform the Measurement Groups into Observations and write to the FHIR Server.
-            //         Exceptions are currently handled in ProcessMeasurementGroups
-            await ProcessMeasurementGroups(measurementGroups, template, log, measurementToEventMapping).ConfigureAwait(false);
+            // Transform the Measurement Groups into Observations and write to the FHIR Server.
+            // Exceptions are currently handled in SendMeasurementGroups
+            await SendMeasurementGroups(measurementGroups, template, log, measurementToEventMapping).ConfigureAwait(false);
+        }
+
+        private async Task ProcessMeasurementGroupEvents(IEnumerable<IEventMessage> events, ILookupTemplate<IFhirTemplate> template, ITelemetryLogger log)
+        {
+            if (events.Count() == 0)
+            {
+                return;
+            }
+
+            foreach (var evt in events)
+            {
+                try
+                {
+                    IEnumerable<MeasurementGroup> groupedMeasurements;
+
+                    try
+                    {
+                        var partitionId = events.FirstOrDefault()?.PartitionId;
+
+                        // decompress the measurement group if it is compressed
+                        byte[] bytes;
+                        if (evt.BodyContentType == "application/gzip")

```

Comment:
If we go with my previous suggestions we could avoid extra object allocations here with something like:

```.net
                        IEnumerable<Measurement> measurementGroup = null;

                        if (evt.BodyContentType == "application/gzip")
                        {
                            using var stream = Compression.DecompressWithGzip(evt.Body.AsStream()); 
                            measurementGroup = System.Text.Json.JsonSerializer.Deserialize<IEnumerable<Measurement>>(stream);
                        }
                        else
                        {
                            using var stream = evt.Body.AsStream();
                            measurementGroup = System.Text.Json.JsonSerializer.Deserialize<IEnumerable<Measurement>>(stream)
                        }
```